src/java.base/share/classes/java/io/Console.java

```diff
@@ -595,22 +596,32 @@ public int read(char[] cbuf, int offset, int length)
         // Set up JavaIOAccess in SharedSecrets
         SharedSecrets.setJavaIOAccess(new JavaIOAccess() {
             public Console console() {
-                if (istty) {
-                    if (cons == null)
-                        cons = new Console();
-                    return cons;
+                if (cons == null) {
+                    // Try loading providers
+                    var consModName =
+                            GetPropertyAction.privilegedGetProperty("jdk.console",
+                                    JdkConsoleProvider.DEFAULT_PROVIDER);
+                    cons = ServiceLoader.load(JdkConsoleProvider.class).stream()
+                       .map(ServiceLoader.Provider::get)
+                       .filter(jcp -> consModName.equals(jcp.getClass().getModule().getName()))
+                       .map(jcp -> jcp.console(istty))
+                       .filter(Objects::nonNull)
+                       .findAny()
+                       .map(jc -> (Console)new ProxyingConsole(jc))
+                       .orElse(istty ? new Console() : null);
                 }
-                return null;
+                return cons;
             }
 
             public Charset charset() {
-                return CHARSET;
+                return cons.charset();
```

Comment:
What happens when cons is null?

Reply:
Turned out that `JavaIOAccess.charset()` is not used anywhere. Will remove this method. Will also add a test.